pkg/controller/framework/shared_informer.go

```diff
@@ -118,40 +124,34 @@ type deleteNotification struct {
 	oldObj interface{}
 }
 
-func (s *sharedIndexInformer) Run(stopCh <-chan struct{}) {
+func (s *coreInformer) Run(stopCh <-chan struct{}) {
 	defer utilruntime.HandleCrash()
 
-	fifo := cache.NewDeltaFIFO(cache.MetaNamespaceKeyFunc, nil, s.indexer)
-
-	cfg := &Config{
-		Queue:            fifo,
-		ListerWatcher:    s.listerWatcher,
-		ObjectType:       s.objectType,
-		FullResyncPeriod: s.fullResyncPeriod,
-		RetryOnError:     false,
-
-		Process: s.HandleDeltas,
+	s.startedLock.Lock()
+	s.started = true
+	s.startedLock.Unlock()
+
+	// This is for backward compatibility of SharedInformer. Currently SharedInformer can have blocking handlers.
+	// Thus, when SharedInformer.AddEventHandler() is called, we automatically converts it to handler that have
+	// background goroutine to queue events.
+	// In the future, we should encourage user to use queuedEventHandler and run it outside informer.
+	// Informer shouldn't do the job of starting goroutines for any handler.
+	for _, h := range s.handlers {
```

Comment:
How about building the `SharedIndexInformer` on top of your `CoreInformer` with a properly cast reference to the `queuedEventHandler`?  That cleanly layers and avoids this.